### PR TITLE
Network: ACL API, schema and CLI

### PR DESCRIPTION
**Pull request by tomponline**
Adds the following new commands to the CLI:

```
lxc network acl list
lxc network acl create <name>
lxc network acl show <name>
lxc network acl edit <name>
lxc network acl rename <name> <new name>
lxc network acl delete <name>
lxc network acl get <name> <key>
lxc network acl set <name> <key> <value>
lxc network acl unset <name> <key>
lxc network acl rule add <name> <direction> [key/value…]
lxc network acl rule remove <name> <direction> [key/value…] [--force]
```

### CODE REVIEW

**Comment by tomponline** [comment]
@stgraber the schema, API and associated CLI are done now (I think!), however I need to add the API docs and some jenkins tests for the CLI functionality. But ready for some review.

File: shared/api/network_acl.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package api
```
Comment (stgraber) [inline shared/api/network_acl.go:1]: Aren't the field descriptions going to be redundant with what we'll have in `doc/`?

If so, we may want to strip them from `shared/api` just to avoid them getting out of sync as we add new supported values down the line.

Eventually we'll want to handle all this stuff through #7874 but until then, best to not duplicate.

Comment (tomponline) [reply shared/api/network_acl.go:1]: fixed

File: client/interfaces.go [hunk 2]
```diff
@@ -228,6 +228,8 @@ type InstanceServer interface {
 	DeleteNetwork(name string) (err error)
 
 	// Network ACL functions ("network_acl" API extension)
+	GetNetworkACLs() (acls []api.NetworkACL, err error)
+	GetNetworkACL(name string) (acl *api.NetworkACL, ETag string, err error)
```
Comment (stgraber) [inline client/interfaces.go:232]: This one was meant for the next commit I think.

File: client/interfaces.go [hunk 3]
```diff
@@ -228,7 +228,7 @@ type InstanceServer interface {
 	DeleteNetwork(name string) (err error)
 
 	// Network ACL functions ("network_acl" API extension)
-	GetNetworkACLs() (acls []api.NetworkACL, err error)
+	GetNetworkACLs() (acl []api.NetworkACL, err error)
```
Comment (stgraber) [inline client/interfaces.go:231]: Why this change? `acls` sounds correct given it's a list, also wrong commit :)

File: client/lxd_network_acls.go [hunk 4]
```diff
@@ -6,7 +6,24 @@ import (
 	"github.com/lxc/lxd/shared/api"
 )
 
-// CreateNetworkACL defines a new network ACL using the provided struct.
+// GetNetworkACLs returns a list of Network ACL structs.
+func (r *ProtocolLXD) GetNetworkACLs() ([]api.NetworkACL, error) {
```
Comment (stgraber) [inline client/lxd_network_acls.go:11]: We'll want a `GetNetworkACLsNames` too for consistency.

File: lxc/network_acl.go [hunk 5]
```diff
@@ -398,3 +403,130 @@ func (c *cmdNetworkACLUnset) Run(cmd *cobra.Command, args []string) error {
...
+### egress: []
+### ingress:
+### - action: accept
+###  source: ""
+###  destination: ""
+###  protocol: ""
+###  sourceport: ""
+###  destinationport: ""
+###  icmptype: ""
+###  icmpcode: ""
```
Comment (stgraber) [inline lxc/network_acl.go:456]: Indent looks off here

**Review by stgraber — CHANGES_REQUESTED** [review]
Just a few things here and there.

**Comment by tomponline** [comment]
@stgraber thanks will sort those, I missed a few references to "networks" earlier and looks like I messed up some rebasing.

File: client/lxd_network_acls.go (hunk 4 above)
Comment (tomponline) [reply client/lxd_network_acls.go:11]: will do.

File: client/interfaces.go (hunk 3 above)
Comment (tomponline) [reply client/interfaces.go:231]: Agreed, fixed.

File: client/interfaces.go (hunk 2 above)
Comment (tomponline) [reply client/interfaces.go:232]: Fixed.

File: lxc/network_acl.go (hunk 5 above)
Comment (tomponline) [reply lxc/network_acl.go:456]: fixed

**Comment by tomponline** [comment]
@stgraber have updated with:

- Docs
- GetNetworkACLNames client function
- PATCH method endpoint
- Rule validation

**Comment by tomponline** [comment]
@stgraber a question on the PATCH method, I've made it so it merges the `config` property with what is already in the database, but for all the other non-config fields (i.e description, ingress, egress) these will replace the existing fields even if empty (which is how the network PATCH endpoint works now). Is that OK?

File: lxd/db/cluster/schema.go [hunk 6]
```diff
@@ -285,6 +285,23 @@ CREATE TABLE "networks" (
     type INTEGER NOT NULL DEFAULT 0,
     UNIQUE (project_id, name)
 );
+CREATE TABLE networks_acls (
+    id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL,
+    project_id INTEGER NOT NULL,
```
Comment (stgraber) [inline lxd/db/cluster/schema.go:290]: Shouldn't we have a foreign key to the project table?

**Comment by stgraber** [comment]
All good except for the schema issue, with that fixed, ready to merge.

File: lxd/db/cluster/schema.go (hunk 6 above)
Comment (tomponline) [reply lxd/db/cluster/schema.go:290]: @stgraber yep I have added that now. The reason I didn't originally was that was was basing the table structure on the `networks` table, which doesn't currently have a foreign key on `project_id`. However expanding my search and looking at the `instances` table showed that there was a foreign key on there, so it looks like I missed this when adding project support for networks.

**Comment by tomponline** [comment]
@stgraber shall I put up a separate PR to add a foreign key for networks table on project_id?

**Comment by stgraber** [comment]
Yeah, that'd be good

**Comment by tomponline** [comment]
@stgraber that should fix the unit test schema errors.
